### PR TITLE
8320995: RISC-V: C2 PopCountVI

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this patch?
Both auto-vect and vector api depends on this intrinsic.
Thanks!

## Performance
Not performance test was done, as this depends on vcpop.v instruction in zvbb extension and the code seqeunce is rather simple than non-intrinsic version.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8320995](https://bugs.openjdk.org/browse/JDK-8320995): RISC-V: C2 PopCountVI (**Sub-task** - P4)
 * [JDK-8320996](https://bugs.openjdk.org/browse/JDK-8320996): RISC-V: C2 PopCountVL (**Sub-task** - P4)

### Reviewers
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) ⚠️ Review applies to [cbfde208](https://git.openjdk.org/jdk/pull/19065/files/cbfde2083229b761e8e47f18ac44743c33b73b2f)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19065/head:pull/19065` \
`$ git checkout pull/19065`

Update a local copy of the PR: \
`$ git checkout pull/19065` \
`$ git pull https://git.openjdk.org/jdk.git pull/19065/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19065`

View PR using the GUI difftool: \
`$ git pr show -t 19065`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19065.diff">https://git.openjdk.org/jdk/pull/19065.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19065#issuecomment-2090750198)

### CODE REVIEW

File: src/hotspot/cpu/riscv/globals_riscv.hpp [hunk 1]
```diff
@@ -115,6 +115,7 @@ define_pd_global(intx, InlineSmallCode,          1000);
   product(bool, UseZtso, false, EXPERIMENTAL, "Assume Ztso memory model")        \
   product(bool, UseZihintpause, false, EXPERIMENTAL,                             \
           "Use Zihintpause instructions")                                        \
+  product(bool, UseZvbb, false, "Use Zvbb instructions")                         \
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/globals_riscv.hpp:118]: Shouldn't this be marked `EXPERIMENTAL` as we have no hardware to test it on?

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 2]
```diff
@@ -3784,6 +3787,37 @@ instruct vconvF2HF(vReg dst, vReg src, vReg vtmp, vRegMask_V0 v0, iRegINoSp tmp)
   ins_pipe(pipe_slow);
 %}
 
+
+// ------------------------------ Popcount vector ------------------------------
+
+instruct vpopcount_mask(vReg dst_src, vRegMask_V0 v0) %{
+  match(Set dst_src (PopCountVI dst_src v0));
+  match(Set dst_src (PopCountVL dst_src v0));
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv_v.ad:3795]: Is there a reason to force input & output being the same vector register?

File: src/hotspot/os_cpu/linux_riscv/riscv_hwprobe.cpp [hunk 3]
```diff
@@ -175,6 +175,9 @@ void RiscvHwprobe::add_features_from_query_result() {
   if (is_set(RISCV_HWPROBE_KEY_IMA_EXT_0, RISCV_HWPROBE_EXT_ZBS)) {
     VM_Version::ext_Zbs.enable_feature();
   }
+  if (is_set(RISCV_HWPROBE_KEY_IMA_EXT_0, RISCV_HWPROBE_EXT_ZVBB)) {
+    VM_Version::ext_Zvbb.enable_feature();
+  }
```
Comment (RealFYang) [inline src/hotspot/os_cpu/linux_riscv/riscv_hwprobe.cpp:180]: I don't think it's appropriate to auto-enable an experimental extension.

File: src/hotspot/cpu/riscv/riscv_v.ad (hunk 2 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/riscv_v.ad]: Seems not, I modify it to usual pattern.

File: src/hotspot/os_cpu/linux_riscv/riscv_hwprobe.cpp (hunk 3 above)
Comment (Hamlin-Li) [reply src/hotspot/os_cpu/linux_riscv/riscv_hwprobe.cpp]: Thanks for catching, fixed.

**Review by luhenry — APPROVED** [review]

**Review by RealFYang — CHANGES_REQUESTED** [review]
One minor comment remains, otherwise looks good. Thanks.

File: test/hotspot/jtreg/compiler/vectorization/TestPopCountVectorLong.java [hunk 4]
```diff
@@ -26,12 +26,14 @@
 * @summary Test vectorization of popcount for Long
 * @requires vm.compiler2.enabled
 * @requires ((os.arch=="x86" | os.arch=="i386" | os.arch=="amd64" | os.arch=="x86_64") & vm.cpu.features ~= ".*avx512bw.*") |
-*           os.simpleArch == "aarch64"
+*           os.simpleArch == "aarch64" |
+*           (os.arch == "riscv64" & vm.cpu.features ~= ".*zvbb,.*")
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply test/hotspot/jtreg/compiler/vectorization/TestPopCountVectorLong.java]: Thanks, fixed.

**Review by RealFYang — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
Thanks @luhenry @RealFYang for your reviewing.

/integrate

**Pull request closed without merging**